internal/p2p/client/consumer.go

```diff
@@ -0,0 +1,127 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/tendermint/tendermint/internal/p2p"
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+var (
+	ErrRequestIDAttributeRequired  = errors.New("envelope requestID attribute is required")
+	ErrResponseIDAttributeRequired = errors.New("envelope responseID attribute is required")
+)
+
+type (
+	// ConsumerHandler is the interface that wraps a Handler method.
+	// This interface must be implemented by the p2p message handler
+	// and must be used in conjunction with the p2p consumer.
+	ConsumerHandler interface {
+		Handle(ctx context.Context, client *Client, envelope *p2p.Envelope) error
+	}
+	// ConsumerMiddlewareFunc is used to wrap ConsumerHandler to provide the ability to do something
+	// before or after the handler execution
+	ConsumerMiddlewareFunc func(next ConsumerHandler) ConsumerHandler
+	// ConsumerParams is p2p handler parameters set
+	ConsumerParams struct {
+		ReadChannels []p2p.ChannelID
+		Handler      ConsumerHandler
+	}
+	recoveryP2PMessageHandler struct {
+		logger log.Logger
+		next   ConsumerHandler
+	}
+	loggerP2PMessageHandler struct {
+		logger log.Logger
+		next   ConsumerHandler
+	}
+	validateMessageHandler struct {
+		allowedChannelIDs map[p2p.ChannelID]struct{}
+		next              ConsumerHandler
+	}
+)
+
+// WithRecoveryMiddleware creates panic recovery middleware
+func WithRecoveryMiddleware(logger log.Logger) ConsumerMiddlewareFunc {
+	hd := &recoveryP2PMessageHandler{logger: logger}
+	return func(next ConsumerHandler) ConsumerHandler {
+		hd.next = next
+		return hd
+	}
+}
+
+// WithLoggerMiddleware creates error logging middleware
+func WithLoggerMiddleware(logger log.Logger) ConsumerMiddlewareFunc {
+	hd := &loggerP2PMessageHandler{logger: logger}
+	return func(next ConsumerHandler) ConsumerHandler {
+		hd.next = next
+		return hd
+	}
+}
+
+// WithValidateMessageHandler creates message validation middleware
+func WithValidateMessageHandler(allowedChannelIDs []p2p.ChannelID) ConsumerMiddlewareFunc {
+	hd := &validateMessageHandler{
+		allowedChannelIDs: map[p2p.ChannelID]struct{}{},
+	}
+	for _, chanID := range allowedChannelIDs {
+		hd.allowedChannelIDs[chanID] = struct{}{}
+	}
+	return func(next ConsumerHandler) ConsumerHandler {
+		hd.next = next
+		return hd
+	}
+}
+
+// HandlerWithMiddlewares is a function that wraps a handler in middlewares
+func HandlerWithMiddlewares(handler ConsumerHandler, mws ...ConsumerMiddlewareFunc) ConsumerHandler {
+	for _, mw := range mws {
+		handler = mw(handler)
+	}
+	return handler
+}
+
+// Handle recovers from panic if a panic happens
+func (h *recoveryP2PMessageHandler) Handle(ctx context.Context, client *Client, envelope *p2p.Envelope) (err error) {
+	defer func() {
+		if e := recover(); e != nil {
```

Comment:
I share this point of view with you.
this middleware was added because of the original implementation has this `recovery` code
https://github.com/dashpay/tenderdash/blob/v0.11-dev/internal/blocksync/reactor.go#L225
this middleware is optional, so we can use it for this handler and omit it for another.